File: Scientific_Journal_Trend_Tracker_Backend/src/models/Journal.ts
import mongoose, { Schema, Document } from "mongoose";

export interface IJournal extends Document {
  name: string;
  issn: string;
  publisher: string;
  impactFactor: number;
  hIndex: number;
  paperCount: number;
  fieldDomain: string;
  isTracked: boolean;
  source: string;
  externalId: string;
  lastSyncedAt?: Date;
}

const journalSchema = new Schema<IJournal>(
  {
    name: {
      type: String,
      required: true,
      unique: true,
    },
    issn: {
      type: String,
      unique: true,
      sparse: true,
    },
    publisher: String,
    impactFactor: Number,
    hIndex: Number,
    paperCount: {
      type: Number,
      default: 0,
    },
    fieldDomain: String,
    isTracked: {
      type: Boolean,
      default: true,
    },
    source: String,
    externalId: String,
    lastSyncedAt: Date,
  },
  {
    timestamps: true,
  },
);

export default mongoose.model<IJournal>("Journal", journalSchema);

File: scientific_journal_trend_tracker_frontend_flutter_app/lib/core/models/journal.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'journal.freezed.dart';
part 'journal.g.dart';

@freezed
abstract class Journal with _$Journal {
  const factory Journal({
    @JsonKey(name: '_id') required String id,
    required String name,
    String? issn,
    String? publisher,
    double? impactFactor,
    int? hIndex,
    @Default(0) int paperCount,
    String? fieldDomain,
    @Default(false) bool isTracked,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) = _Journal;

  factory Journal.fromJson(Map<String, dynamic> json) => _$JournalFromJson(json);
}
